minitests/iostandard/features/generate.py

```diff
@@ -0,0 +1,491 @@
+
+IOBUF_NOT_ALLOWED = [
+    'HSTL_I',
+    'HSTL_I_18',
+    'SSTL18_I',
+]
+
+DIFF_MAP = {
+    'SSTL135': 'DIFF_SSTL135',
+}
+
+
+def gen_iosettings():
+    """
+    A generator function which yields all possible IO settings combintions.
+    """
+
+    IOSTANDARDS = (
+        'LVCMOS12',
+        'LVCMOS15',
+        'LVCMOS18',
+        'LVCMOS25',
+        'LVCMOS33',
+        'LVTTL',
+        'SSTL135',
+
+        # Those are available but not currently fuzzed.
+        #        'SSTL135_R',
+        #        'SSTL15',
+        #        'SSTL15_R',
+        #        'SSTL18_I',
+        #        'SSTL18_II',
+        #        'HSTL_I',
+        #        'HSTL_I_18',
+        #        'HSTL_II',
+        #        'HSTL_II_18',
+        #        'HSUL_12',
+        #        'MOBILE_DDR',
+    )
+
+    DRIVES = defaultdict(lambda: [None])
+    DRIVES.update(
+        {
+            "LVTTL": [4, 8, 12, 16, 24],
+            "LVCMOS12": [4, 8, 12],
+            "LVCMOS15": [4, 8, 12, 16],
+            "LVCMOS18": [4, 8, 12, 16, 24],
+            "LVCMOS25": [4, 8, 12, 16],
+            "LVCMOS33": [4, 8, 12, 16],
+        })
+
+    SLEWS = ("SLOW", "FAST")
+
+    for iostandard in IOSTANDARDS:
+
+        # Single ended
+        for drive in DRIVES[iostandard]:
+            for slew in SLEWS:
+                yield {"iostandard": iostandard, "drive": drive, "slew": slew}
+
+        # Differential
+        if iostandard in DIFF_MAP:
+            for drive in DRIVES[iostandard]:
+                for slew in SLEWS:
+                    yield {
+                        "iostandard": DIFF_MAP[iostandard],
+                        "drive": drive,
+                        "slew": slew
+                    }
+
+
+# =============================================================================
+
+
+def run():
+    """
+    Main.
+    """
+
+    # Load IOB data
+    iob_sites = load_iob_sites("iobs-{}.csv".format(os.getenv("VIVADO_PART")))
+
+    # Generate IOB site to package pin map and *M site to *S site map.
+    site_to_pkg_pin = {}
+    master_to_slave = {}
+
+    for region, sites in iob_sites.items():
+        tiles = defaultdict(lambda: {})
+
+        for site in sites:
+            site_to_pkg_pin[site["name"]] = site["pkg_pin"]
+            if site["type"] == "IOB33M":
+                tiles[site["tile"]]["M"] = site
+            if site["type"] == "IOB33S":
+                tiles[site["tile"]]["S"] = site
+
+        for sites in tiles.values():
+            master_to_slave[sites["M"]["name"]] = sites["S"]["name"]
+
+    # Generate designs
+    iosettings_gen = gen_iosettings()
+    design_index = 0
+    while True:
+
+        num_inp = 0
+        num_out = 0
+        num_ino = 0
+
+        # Generate clock regions
+        region_data = []
+        for region in sorted(list(iob_sites.keys())):
+
+            # Get IO settings
+            try:
+                iosettings = next(iosettings_gen)
```

Comment:
Why not use:

`for region, iosettings = sorted(zip(iob_sites, gen_iosettings), key=lambda x: x[0]):`

Reply:
Can you describe in words what you were trying to do with: https://github.com/SymbiFlow/prjxray/pull/1157/files?short_path=02164c0#diff-84790ccfe259a048566f5d1695b477ddR156-R162 ?

Reply:
So there is a generator of IO settings combinations `gen_iosettings()`. But since there is 5 IO banks on the Artix7 part I need to get 5 of those settings per design. And off course the numer of IO settings combinations may not be dividable by 5 so the last design may have less than 5 banks used.

And this is implemented as a while loop that generates design. Inside it there is a loop over clock regions / IO banks which may terminate if we run out of IO setting combinations.